src/wasi_unstable/mod.rs

```diff
@@ -236,134 +254,140 @@ pub const WHENCE_CUR: Whence = __WASI_WHENCE_CUR;
 pub const WHENCE_END: Whence = __WASI_WHENCE_END;
 pub const WHENCE_SET: Whence = __WASI_WHENCE_SET;
 
-pub fn clock_res_get(clock_id: ClockId) -> (Errno, Timestamp) {
+macro_rules! unsafe_wrap0 {
+    {$f:expr} => {
+        unsafe {
+            if let Some(code) = NonZeroU16::new($f) {
+                Err(code)
+            } else {
+                Ok(())
+            }
+        }
+    };
+}
+
+macro_rules! unsafe_wrap1 {
+    {$var:ident, $f:expr,} => {
+        unsafe {
+            if let Some(code) = NonZeroU16::new($f) {
+                Err(code)
+            } else {
+                Ok($var.assume_init())
+            }
+        }
+    };
+}
+
+pub fn clock_res_get(clock_id: ClockId) -> Result<Timestamp, Errno> {
     let mut resolution = MaybeUninit::<Timestamp>::uninit();
-    unsafe {
-        (
-            __wasi_clock_res_get(clock_id, resolution.as_mut_ptr()),
-            resolution.assume_init(),
-        )
+    unsafe_wrap1!{
+        resolution,
+        __wasi_clock_res_get(clock_id, resolution.as_mut_ptr()),
     }
 }
 
-pub fn clock_time_get(clock_id: ClockId, precision: Timestamp) -> (Errno, Timestamp) {
+pub fn clock_time_get(clock_id: ClockId, precision: Timestamp) -> Result<Timestamp, Errno> {
     let mut time = MaybeUninit::<Timestamp>::uninit();
-    unsafe {
-        (
-            __wasi_clock_time_get(clock_id, precision, time.as_mut_ptr()),
-            time.assume_init(),
-        )
+    unsafe_wrap1!{
+        time,
+        __wasi_clock_time_get(clock_id, precision, time.as_mut_ptr()),
     }
```

Comment:
What would you think of reversing the order of the parameters in `unsafe_wrap1`, so it's 
```rust
    unsafe_wrap1!{
        __wasi_clock_time_get(clock_id, precision, time.as_mut_ptr()),
        time,
    }
```
which puts them closer to the order they appear in the expanded form, making it easier to read?

Reply:
Will do. I also thought about writing something like this:
```rust
unsafe_wrap1!{ __wasi_clock_time_get, clock_id, precision, Timestamp }
```
Unwrapped code will handle creating `MaybeUninit` and the `.as_mut_ptr()` part.